From ba71f54fb6f066bb4c2973e58b86b5a5062921d0 Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Tue, 4 Nov 2014 20:42:16 +0800
Subject: [PATCH] MLK-9794-3 usb: common: otg: HNP polling condition update
 for b host

If it's A host and b_hnp_enable is set, or B host, we can do HNP polling
because B host will not check OTG descriptor and set b_hnp_enable flag
when enumeration.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 6148c9df6d1c68dfc91cfe274101058c8edf979b)
---
 drivers/usb/common/usb-otg-fsm.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/common/usb-otg-fsm.c b/drivers/usb/common/usb-otg-fsm.c
index 4600808..0afde68 100644
--- a/drivers/usb/common/usb-otg-fsm.c
+++ b/drivers/usb/common/usb-otg-fsm.c
@@ -432,7 +432,8 @@ int otg_hnp_polling(struct otg_fsm *fsm)
 	enum usb_otg_state state = fsm->otg->state;
 	struct usb_otg_descriptor *desc = NULL;
 
-	if (state != OTG_STATE_A_HOST && state != OTG_STATE_B_HOST)
+	if ((state != OTG_STATE_A_HOST || !fsm->b_hnp_enable) &&
+					state != OTG_STATE_B_HOST)
 		return -EINVAL;
 
 	udev = usb_hub_find_child(fsm->otg->host->root_hub, 1);
-- 
1.8.0

